### ISSUE TITLE
Proposal (RFC): [experimental] Rekor cataloger/certifier

### ISSUE THREAD

**Issue by lumjjb** [guacsec/guac#275]
Along the lines of the [GUAC Beta](https://github.com/guacsec/guac/issues/251) roadmap, we are taking the approach of ingesting a set of software (through SBOMs), and augmenting the graph by expanding based on what we're seeing be ingested. Initially the metadata focused on will be from deps.dev and osv.dev. 

This issue talks about exploring an approach to also think about getting some attestation metadata from [Rekor](https://github.com/sigstore/rekor) as well... **This is not a commitment to implement this, but an issue for discussion.**

**DISCLAIMER:** My understanding is that the rekor search function isn't intended/committed to be part of rekor's core function, and it doesn't necessarily want to be the store of attestations... Ideally rekor would only be used as a verification mechanism, and attestations would be obtained from separate storage mechanisms (OCI registries, package managers, etc.)... Work for that is still planned.. And while the conventions are still being ironed out for various ecosystems, this would be more of a quick intermittent way to let folks see what the effect of all this will be in the long term (with sigstore).

Some of the inspiration for this proposal comes from the work with the rekor cataloger/certifier worked on by @mdeicas implemented as part of [Syft](https://github.com/anchore/syft/), as part of the [Kubecon Talk "SBOM X-ray Superpowers"](https://kccncna2022.sched.com/event/182GT/sbom-x-ray-superpowers-making-better-sboms-using-sboms-brandon-lum-google-chris-phillips-anchore?iframe=no). 

Separate from #68 , this rekor cataloger/certifier will not ingest all entries from a rekor instance but use the heuristic of the [rekor hash search feature ](https://docs.sigstore.dev/rekor/cli/#search). Instead, it will have an input as a list of sha256 hashes... In the longer term, IF this is implemented, it will be deprecated as other collector implementations (like #211, #208 , etc.) would be the source of truth for discovery of attestations.

**Comment by lumjjb** [comment]
Had some thoughts on the plane back.. so wrote some of it down.. will see if it's still a good idea after some sleep 😂

**Comment by lukehinds** [comment]
For your disclaimer, this is largely true. Querying rekor should be for Inclusion proof guarantees. e.g. I have this SLSA / in-toto dsse wrapped attestation in my graph, is it transparent / observable in rekor (is everyone seeing what I am seeing, or have I been singled out and fed something bad). Rekor provides a redis instance for speedy queries, but there are of course none of the guarantees you get with the merkle tree (anyone could tamper with redis).

Having said that, it's a useful store to gather graph data. It's not abusive to monitor rekors index and if it increments, pop off the newest entry, check its type (e.g. `application/vnd.in-toto+json`) and load it into the graph. Anyone then looking at that node, knows it's got the belt and braces guarantees of being in rekor as well.

I can see a UX / architecture where someone generates provenance, pushes to rekor and the graph pulls down the n entry. Rekor is not designed to be a database with optimised queries , joins etc, so that is offloaded to the graph along with all the nice powerful cypher query niceness, but you still then have a statement of observable truth within rekor, that is tamper proof etc.

**Comment by lumjjb** [comment]
Yup - I think that probably be the ideal architecture.. One area that we're struggling with is the storage/retrieval of attestations blobs...

But i believe there's stuff around the pub/sub that @haydentherapper and @asraa shared that should help with this! 

As an aside and just to capture past discussions in the issue,  I chatted about a backpointer with @asraa before, but seems like that has its own set of difficulties when the information is out of date and needs an update/new entry, either as unauthenticated data or issues needing to implement an implicit "log revocation" mechanism.

Probably will leave this issue up for a bit and if there's less interest on the certifier, we'll focus more efforts on fleshing out the direct rekor integration as part of #76 from @rgreinho !

**Comment by haydentherapper** [comment]
I don't think we've written any docs on this feature, but I could imagine we would publish each entry created to a pub/sub queue that listeners could subscribe to. For verifiability, much like if GUAC mirrored Rekor entries, you'd either need:
* GUAC or the pub/sub listener would need to persist a promise of inclusion from Rekor, much like certificates contain promises of inclusion in certificate transparency logs
* A consumer of the mirror or GUAC would query Rekor for a proof of inclusion before trusting a result
